### PR TITLE
[webgpu] Add cumprod and cumsum

### PR DESCRIPTION
**Pull request by axinging**
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/6317)

### CODE REVIEW

**Comment by axinging** [comment]
@qjia7 @haoyunfeix @xhcao @gyagp , PTAL

File: tfjs-backend-webgpu/src/cum_webgpu.ts [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+}
+
+export class CumProgram implements WebGPUProgram {
+  outputShape: number[];
+  shaderKey: string;
+  dispatchLayout: {x: number[]};
+  dispatch: [number, number, number];
+  variableNames = ['x'];
+  workGroupSize: [number, number, number];
+  uniforms = 'index : f32,';
```
Comment (qjia7) [inline tfjs-backend-webgpu/src/cum_webgpu.ts:34]: It seems that index should be a `i32` not a `f32`.

File: tfjs-backend-webgpu/src/cum_webgpu.ts [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+      idxString = (this.reverse ? 'end + pow2' : 'end - pow2');
+    }
+    return `
+     ${getMainHeaderAndGlobalIndexString()}
+       var coords = getCoordsFromIndex(index);
+
+       let end = ${getFinalCoord(rank, 'coords', this.op)};
+       var val = ${val};
+       let pow2 = i32(pow(2.0, uniforms.index));
+       if (index < uniforms.size) {
```
Comment (qjia7) [inline tfjs-backend-webgpu/src/cum_webgpu.ts:79]: Is it better to put the boundary check at the beginning?

Comment (qjia7) [inline tfjs-backend-webgpu/src/cum_webgpu.ts:79]: Is it better to put the boundary check at the beginning?

File: tfjs-backend-webgpu/src/kernels/Cumprod.ts [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+import {backend_util, Cumprod, CumprodAttrs, CumprodInputs, KernelConfig, KernelFunc, TensorInfo} from '@tensorflow/tfjs-core';
+
+import {WebGPUBackend} from '../backend_webgpu';
+import {CumOpType, CumProgram} from '../cum_webgpu';
+
+import {identity} from './Identity';
+import {transpose} from './Transpose';
+
+export function cumprod(
```
Comment (qjia7) [inline tfjs-backend-webgpu/src/kernels/Cumprod.ts:26]: `cumprod` and `cumsum` has many duplicated code. Is it possible to call a common function like `Cum_impl` to reduce the duplication?

Comment (qjia7) [inline tfjs-backend-webgpu/src/kernels/Cumprod.ts:26]: `cumprod` and `cumsum` has many duplicated code. Is it possible to call a common function like `Cum_impl` to reduce the duplication?

File: tfjs-backend-webgpu/src/cum_webgpu.ts (hunk 1 above)
Comment (axinging) [reply tfjs-backend-webgpu/src/cum_webgpu.ts:35]: pow(f32, i32) is not supported.

File: tfjs-backend-webgpu/src/cum_webgpu.ts (hunk 2 above)
Comment (axinging) [reply tfjs-backend-webgpu/src/cum_webgpu.ts:75]: Thanks, done

Comment (axinging) [reply tfjs-backend-webgpu/src/cum_webgpu.ts:75]: Done

File: tfjs-backend-webgpu/src/kernels/Cumprod.ts (hunk 3 above)
Comment (axinging) [reply tfjs-backend-webgpu/src/kernels/Cumprod.ts:24]: Done, thanks!

Comment (axinging) [reply tfjs-backend-webgpu/src/kernels/Cumprod.ts:24]: Done

File: tfjs-backend-webgpu/src/cum_webgpu.ts [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+}
+
+export class CumProgram implements WebGPUProgram {
+  outputShape: number[];
+  shaderKey: string;
+  dispatchLayout: {x: number[]};
+  dispatch: [number, number, number];
+  variableNames = ['x'];
+  workGroupSize: [number, number, number];
+  // pow(f32, i32) is not supported, use pow(f32, f32) instead.
```
Comment (qjia7) [inline tfjs-backend-webgpu/src/cum_webgpu.ts:34]: nit: `pow(i32, i32) is not supported`? Usually, we use the same type.

File: tfjs-backend-webgpu/src/kernels/Cum_impl.ts [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+import {backend_util, TensorInfo} from '@tensorflow/tfjs-core';
+
+import {WebGPUBackend} from '../backend_webgpu';
+import {CumOpType, CumProgram} from '../cum_webgpu';
+
+import {identity} from './Identity';
+import {transpose} from './Transpose';
+
+export function cumpImpl(
```
Comment (qjia7) [inline tfjs-backend-webgpu/src/kernels/Cum_impl.ts:26]: nit: typo `cumpImpl` -> `cumImpl`?

**Review by qjia7 — APPROVED** [review]
LGTM with two nits.

File: tfjs-backend-webgpu/src/cum_webgpu.ts (hunk 4 above)
Comment (axinging) [reply tfjs-backend-webgpu/src/cum_webgpu.ts:34]: Not support.

File: tfjs-backend-webgpu/src/kernels/Cum_impl.ts (hunk 5 above)
Comment (axinging) [reply tfjs-backend-webgpu/src/kernels/Cum_impl.ts]: done

**Pull request merged**